### ISSUE TITLE
quick fix.. 'primary labels'/'priority labels'

### ISSUE THREAD

**Issue by dobkeratops** [bbernhard/imagemonkey-core#215]
what if you could mark one (or more) of the labels on an image as a priority task.. and the random/default mode would serve these before any others (perhaps prefixing with an asterisk in the label list), eg "*car,*road,*bin,building,tree,pavement,person" for an image where the cars and road are easy to pick out, but the pavement is thin , and there's just a few pixels of people in the distance. "*head/person,person" for a closeup crowd scene. easy to pick out the heads first. etc

- the idea would be to mark the most prominent and easy objects (bigger) as such;
- the upload page (where it asks you to pick a label) could assume that is going to be a primary/priority label?

This could perhaps improve the situation of searching for easy tasks - if these labels are the more obvious things in the scenes. 

I would see this as a temporary fix prior to getting a unified change-label-as-you-annotate mode; so the question is : would this extra complication just clutter the system (another thing to explain to users). would it be better to ignore this idea and just focus on a new unified mode and take the time to do it well.

**Comment by dobkeratops** [comment]
> I am not sure if it's worth the effort.

right that's what worried me about suggesting it - know a myriad of small features can themselves become overwhelming.. it might just distract from the big long term feature (unified mode).

> I am wondering, if we can implicitly assume that the first added labels are the most prominent ones? 

that's an interesting idea. I suppose you could still pick images at random, then chose their task from the next one in the given order. That would involve no new UI.

**Comment by dobkeratops** [comment]
Ah another compromise suggestion:-

How about having a default priority set per label.. i.e. the label list itself is ordered.

the system could pick images at random to serve, but within the image present the labelling tasks in an order based on the default priority

I would suggests:-

(i) components (head/person etc) - but never add them automatically. rely on "add labels" mode for users to specify which images are worth doing the components on. I only add these where I'd want to draw boxes around these parts.
(ii) objects like people, cars, items of furniture = high priority
(iii) 'solid' urban ground types  road,pavement
(v) buildings (these are 'urban landscape' elements, or sometimes more like objects dotted in natural landscapes)
(iv) natural features (river, sea,mountain,..)
(v) trees - almost always extremely difficult to do, so always make these low (.. so users are unlikely to be asked to label them.. but they an always opt in via the search mode)

The next step would be to have a different priority list based on an overall judgement of the image (interior vs exterior, rural vs urban, workplace vs home). Small vs large objects.. would vary depending on the scene. if you were in a street scene, plates on a table infront of a cafe would be fiddly, whereas if it was a closeup of a table, those would be higher priority..  for the moment I would suggest just lumping large and small objects together as priority (ii)

**Comment by bbernhard** [comment]
Interesting idea!

Not sure if I understood it correctly, so please correct me, if I am wrong. 

* each label has a pre-defined priority (defined in the `labels.json`?)
* if a user selects an image for annotation, the system picks the annotation tasks with the easy labels first
* in order to improve the "pick easy tasks" algorithm, we also look at the scene, object size...

I am wondering whether a neural net can help us with the prioritization here? Otherwise, I imagine it will be pretty hard to give a good priority estimation, as there are a lot of things to consider (object size, scene, image quality, image size, ...). Personally, e.q I am totally fine with doing pixel based annotations of complex objects, as soon as the object is large enough and the image quality is good. 

Not sure if that actually works, but what we could try is: We could take a pre-trained Mask-RCNN model that was trained on a lot of different labels and let our images run through Mask-RCNN. Mask RCNN segments our image and (hopefully) finds a few shapes. The shapes are probably not pixel perfect, but hopefully not that far off the real thing. If the found shape size is >= half of the image size, we mark it as easy to annotate.

**Comment by dobkeratops** [comment]
> * if a user selects an image for annotation, the system picks the annotation tasks with the easy labels first

my idea was to pick the images (with remaining tasks) at random, but then choose the label within each image based on the priority; the hope is this would still give a range of tasks, just tending to bias toward the easy ones - and there would still be exceptions to the rule (if there are fewer labels in an image, overall they might be easier examples?)

>> * in order to improve the "pick easy tasks" algorithm, we also look at the scene, object size...

that could work for components perhaps? - e.g once you've got a large 'dog' annotated then it might be safe to automatically add it's peices as new tasks?
